[Spring 경로 조회 - 2단계] 포키(권예진) 미션 제출합니다.

YJGwon opened the pull request:
안녕하세요 카일!
요구사항이 추가되면서 구조를 어떻게 가져갈지에 대한 고민이 많았어요.
기존의 제 코드에서는 Line이 Sections를 가지고 있는 형태였기에, 경로 조회 시 거쳐가는 노선의 정보를 어떻게 조회할 지가 관건이었습니다.

결론적으로는 저는 `노선/구간/역 목록 관리`와 `경로 조회`를 다른 문제 영역이라고 인식했어요.
따라서 경로 조회는 `Lines`나 `Sections`와는 관련 없이 별개의 도메인 객체들(`JGrapht`나, `PathEdge`, `Path` 등)이 일하도록 구현했습니다.
새로운 요구사항 추가에 따라 기존의 요구사항에 맞게 잘 협력하고 있던 객체들의 구조를 너무 많이 변경해야 한다면

- 첫 번째로 설계에 대한 의심을 해봐야겠지만
- **새로운 요구사항이 과연 기존 요구사항과 같은 영역의 문제가 맞는지**도 의심해야 하지 않을까

하는 생각을 하게 되었어요.
다른 크루들과 이야기해봐도 저와 같은 생각을 하는 크루들은 별로 없고 대부분 Section과 Line의 조합법을 고민하고 있더라고요 😓 너무 엉뚱하고 이상한 생각일 수 있지만, 그렇기에 더더욱 피드백을 받아보고 싶어서 제 생각대로 구현해봤습니다.

지난 단계에서 merge하며 주신 comment들도 최대한 제 생각을 녹여서 반영하려 노력해보았는데, 부족한 점 많이 많이 알려주시면 감사하게 학습해보겠습니다!! 이번 단계 PR도 많이 늦어서 죄송합니다 😭 

---
### 서비스 -> Dao VS 서비스 -> 다른 서비스

저는 Service가 서로를 의존하지 않게 구현했는데요!
Service와 Dao가 1:1로만 의존하는 수직적인 구조는 유연하지 못하다고 생각했기 때문이에요.
특히나 저는 Domain 설계를 최대한 DB table 구조에 영향 받지 않게 하려 했기 때문에, 한 Service가 하나의 Dao만 참조하게 하는 것이 답답하게 느껴졌습니다.
**Service class는 도메인 객체를 기준으로 작성하고, 각 service에서 여러 repository들을 사용해서 자유롭게 데이터를 가져오고 조합**할 수 있는 것이 좋게 느껴졌어요. 수평 계층을 잘 분리해두었다면 충분히 누릴 수 있는 장점 아닌가? 라고 생각했구요.
그런데 오늘 피드백 강의에서 이런 [Discussion](https://github.com/woowacourse/retrospective/discussions/15)이 열렸고, 다른 크루들의 생각을 읽어보면서 혼란이 왔습니다 😱

- Service에서 여러 Domain을 사용하게 된다면 각 Service의 책임은 너무 무거운 것이 아니냐
- Service간 순환 참조 문제가 발생한다면 단순히 의존을 다르게 하는 것이 아니라 Service의 설계 자체를 의심해야 한다
- Dao만을 의존하게 하면 Service에서 이루어지는 검증 등의 비즈니스 로직이 중복될 수 있다

와 같은 의견들이 인상적이었어요.
그러나 아직은 제가 큰 규모의 Application을 구현해본 경험도, Layered Architecture에 대한 지식도 많이 부족해서 완전히 이해되지는 않아요..! ㅠㅠ 그래서 이번 미션 까지는 Dao를 직접 의존하는 방식을 유지했습니다.

YJGwon committed:
docs(docs/README): 예외 사항 및 기능 목록 추가

YJGwon committed:
feat(JdbcStationDao): 존재하지 않는 id로 조회 시 예외 발생

- id값을 받는 method의 Parameter type 모두 Long으로 통일

YJGwon committed:
refactor(ui/controller): 성공 응답 body를 argument로 변경

YJGwon committed:
refactor(Sections, PathSearcher): return하는 collection 객체 불변처리

YJGwon committed:
refactor(Path): 역 목록과 거리 가지는 Path 분리

- ShortestPathFactory: 주어진 역과 구간 목록으로 jgrapht의 DijkstraShortestPath 생성
- 존재하지 않는 역일 시 예외 전환(message localize)

YJGwon committed:
test(ExtraFareTest): edge case 추가

YJGwon committed:
feat(Distance): 음수로 생성 시 예외 발생

- ExtraFare가 요금 계산 로직을 Functional Interface로 반환: Distance의 value에 대한 관리 책임을 Distance에게 온전히 주기 위함

YJGwon committed:
feat(Path): 예외 처리 추가

- 출발역과 도착역이 같을 때
- 구간을 찾을 수 없을 때

YJGwon committed:
refactor(ExtraFare): 조건에 따른 추가 요금을 각 Enum 객체에서 계산, 분기문 제거

YJGwon committed:
refactor(ShortestPathFactory): stations를 따로 load하지 않고 sections에서 추출

YJGwon committed:
feat(PathEdge): custom WeightedEdge 구현

YJGwon committed:
feat(Path): custom WeightedEdge 구현

YJGwon committed:
feat: 노선 등록 시 추가 요금 함께 등록

- Fare 추가: 요금 입력값에 대한 검증

YJGwon committed:
feat: 노선 조회 시 추가 요금 함께 조회

YJGwon committed:
feat: 노선 수정 시 추가 요금 함께 수정

YJGwon committed:
feat: 경로 이용 요금에 노선별 추가 요금 부과

YJGwon committed:
refactor: 거리 요금 계산 결과 Fare 객체로 포장

YJGwon committed:
refactor(subway/domain): package 분리

YJGwon committed:
feat: 나이에 따른 요금 할인 적용

YJGwon committed:
test(PathAcceptanceTest): Fixture method 활용

kang-hyungu commented on src/main/java/wooteco/subway/dao/JdbcLineDao.java:
```diff
@@ -33,9 +33,13 @@ public JdbcLineDao(DataSource dataSource, JdbcTemplate jdbcTemplate) {
+    public Line save(Line line, Fare extraFare) {
         try {
-            SqlParameterSource param = new BeanPropertySqlParameterSource(line);
+            Map<String, Object> param = Map.of(
```
try catch 문에서 param 변수를 설정할 필요는 없어보이네요.

kang-hyungu commented on src/main/java/wooteco/subway/dao/JdbcLineDao.java:
```diff
@@ -33,9 +33,13 @@ public JdbcLineDao(DataSource dataSource, JdbcTemplate jdbcTemplate) {
+                    "extra_fare", extraFare.getValue()
+            );
             final Long id = jdbcInsert.executeAndReturnKey(param).longValue();
             return createNewObject(line, id);
```
createNewObject 메서드는 제거하시고 Line 객체를 새로 생성해서 반환하도록 만드세요
리플렉션을 쓸 이유가 없습니다.

kang-hyungu commented on src/main/java/wooteco/subway/dao/JdbcLineDao.java:
```diff
@@ -66,6 +70,17 @@ public Line findById(Long id) {
+        final String sql = "SELECT extra_fare FROM line WHERE id = ?";
+        try {
+            Integer extraFare = jdbcTemplate.queryForObject(sql, Integer.class, id);
+            return new Fare(extraFare);
```
이런 경고가 뜨는데 해결 할 방법 없을까요?

<img width="654" alt="image" src="https://user-images.githubusercontent.com/15669435/170614804-ed6fdac8-d2aa-42e4-b2e5-16bf625b581c.png">

kang-hyungu commented on src/main/java/wooteco/subway/dao/JdbcSectionDao.java:
```diff
@@ -36,19 +37,9 @@ public void save(Section section, Long lineId) {
     @Override
     public void save(Sections sections, Long lineId) {
-        LinkedList<Section> values = sections.getSections();
+        List<Section> values = sections.getSections();
```
values라는 변수명은 코드를 파악하기 어렵게 만듭니다.
일급 컬렉션 같은 특정 상황이 아니면 의미 있는 이름을 부여해주세요.

```suggestion
        List<Section> sections = sections.getSections();
```

kang-hyungu commented on src/main/java/wooteco/subway/dao/JdbcSectionDao.java:
```diff
@@ -36,19 +37,9 @@ public void save(Section section, Long lineId) {
-    }
-
     @Override
     public void save(Sections sections, Long lineId) {
```
이 메서드는 최소 놀람 원칙을 위배하고 있습니다.
일반적인 경우에 save라는 메서드명을 보면 db에 데이터를 저장하겠구나라고 생각할텐데 실제 동작은 update를 하고 있습니다.
save 또는 update는 서비스 클래스에서 결정하도록 만들어주세요
dao는 crud 처리에 집중하도록 설계해야 코드 복잡도를 낮출 수 있습니다.

YJGwon replied:
Sections같은 경우 한 section을 추가/삭제한 뒤 어떻게 변해있을지 그 경우의 수가 많은데요,
기존의 구간이 축소되었을지, 병합되었을지, 몇 개의 index가 변했을지를 Sections객체 외부에 알릴 필요가 없게 만들고 싶었어요.
service에서 update와 insert를 결정하게 하려면 Sections의 각 section 중 어떤 것이 변경되었고 추가되었는지 service에 알려야 할텐데 저는 굳이 그래야 하나 라는 생각이 듭니다.
Sections의 상태 변경은 모두 도메인에게 맡기고, service입장에서는 뭐가 얼마나 어떻게 바뀌었는지는 모른 채 Dao에게 `이 대로 저장해줘~`라고 요청을 보내도록 하는 것이 저의 의도였습니다.

kang-hyungu commented on src/main/java/wooteco/subway/dao/JdbcStationDao.java:
```diff
@@ -57,10 +58,14 @@ public List<Station> findAll() {
     @Override
     public Station findById(Long id) {
         final String sql = "SELECT * FROM station WHERE id = ?";
```
[SELECT * 쓰지 말자](https://hyewoncc.github.io/2022/05/25/sql-no-wildcard.html)

kang-hyungu commented on src/main/java/wooteco/subway/dao/SectionDao.java:
```diff
@@ -15,4 +16,6 @@ public interface SectionDao {
 
     List<Section> findAll();
+
+    Fare findExtraFareById(Long id);
```
왜 SectionDao에서 처리하나요?
db 스키마에는 line에 extra_fare로 저장하고 있는데 실제 객체 관계에서는 line과 fare가 관련이 없네요
fare가 section과 연결되어 있지도 않은데 SectionDao에서 조회하는게 맞을까요?

YJGwon replied:
도메인 객체가 꼭 DB 스키마와 같은 구조를 가질 필요는 없다고 생각합니다.
DB의 구조는 해당 table이 나타내는 개체의 `속성`들이 column이 되지만
Domain 객체의 field 구조는 해당 객체가 어떤 일을 할 것이고 그 일을 할 때 어떤 값이 필요하며 어떤 `상태`를 지니게 되는지에 따라 구성해야 한다고 생각해요.
따라서 DB의 설계에 상관없이 Domain 관계를 설정했습니다.
DB에서 어떤 값을 어떻게 조립해서 가져올지에 대해 좀 더 자유롭게 생각할 수 있어야
- DB는 데이터 저장, 관리 관점에서 효율적인 형태로
- Domain은 협력 관점에서 효율적인 형태로

구성할 수 있지 않나 하는 것이 지금의 생각입니다.

kang-hyungu commented on src/main/java/wooteco/subway/domain/path/Age.java:
```diff
@@ -0,0 +1,24 @@
+        }
+    }
+
+    boolean isSameOrBiggerThan(int age) {
```
1. 의도한 대로 동작하는지 테스트 코드로 검증해보시기 바랍니다.
2. 도메인 클래스의 모든 public 메서드는 테스트 코드로 검증해야 변경이 생겼을 때 영향 받는 객체가 없는지 빠르게 파악할 수 있습니다.

kang-hyungu commented on src/main/java/wooteco/subway/domain/path/AgeFare.java:
```diff
@@ -0,0 +1,35 @@
+import java.util.function.Predicate;
+
+public enum AgeFare {
+    KID((age) -> age.isSameOrBiggerThan(6) && age.isSmallerThan(13),new Fare(350), 0.5),
```
```suggestion
    KID(age -> age.isSameOrBiggerThan(6) && age.isSmallerThan(13), new Fare(350), 0.5),
```

kang-hyungu commented on src/main/java/wooteco/subway/domain/path/Fare.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    @Override
+    public int compareTo(Fare otherFare) {
```
https://rules.sonarsource.com/java/RSPEC-1210

이펙티브 자바 3판 
아이템 14. COMPARABLE을 구현할지 고려하라도 참고해보세요

kang-hyungu commented on src/main/java/wooteco/subway/domain/path/Path.java:
```diff
@@ -0,0 +1,73 @@
+            Fare extraFare = path.getEdgeList().stream()
+                    .map(PathEdge::getExtraFare)
+                    .max(Fare::compareTo)
+                    .get();
```
optional은 언제 쓰는걸까요?
get을 썼을때 문제는 없을까요?
orElse를 적용해보시기 바랍니다.

kang-hyungu commented on src/main/java/wooteco/subway/domain/path/ShortestPathFactory.java:
```diff
@@ -0,0 +1,40 @@
+import wooteco.subway.domain.Station;
+import wooteco.subway.domain.section.Section;
+
+public class ShortestPathFactory {
```
팩터리 메서드로 만들기보단 최소거리를 계산하는 책임을 맡기는게 낫지 않을까요?
PathService에서 추가로 처리해야되는 작업이 많이 보이고 메서드가 10줄이 넘어가게 되네요

kang-hyungu commented on src/main/java/wooteco/subway/domain/path/ShortestPathFactory.java:
```diff
@@ -0,0 +1,40 @@
+
+public class ShortestPathFactory {
+
+    public static DijkstraShortestPath<Station, PathEdge> getFrom(Map<Section, Fare> edges) {
```
리턴 타입에 특정 라이브러리의 클래스가 반환되지 않도록 만드세요
jgraph를 사용하지 않도록 요구사항이 바뀌면 영향 받는 코드가 많아집니다.

kang-hyungu reviewed (changes requested):
포키 몇 가지 코멘트 남겼어요
수정해보시고 궁금한 점 생기면 코멘트 남기시거나 오프라인으로 질문주셔도 좋습니다.

KIMSIYOUNG commented on src/main/java/wooteco/subway/domain/path/Path.java:
```diff
@@ -0,0 +1,73 @@
+import java.util.List;
+import java.util.Objects;
+import org.jgrapht.GraphPath;
+import org.jgrapht.alg.interfaces.ShortestPathAlgorithm;
```
1단계에서 드렸던 리뷰에 대해서 생각해보셨나요? 이 부분에 대해서 생각 공유해주시면 좋을 것 같아요.

https://github.com/woowacourse/atdd-subway-path/pull/247#discussion_r878781871

YJGwon replied:
처음 해당 코멘트를 받고 나서는 외부 Library역시 하나의 도메인 객체로 보고 Service에서 협력시키는 방향을 생각했어요.
그러나 여전히 `이 요구사항에 맞게 Library를 어떻게 사용할지도 하나의 도메인이지 않나?`하는 생각이 혼재되어있었고, 그 혼란이 2단계 제출 코드에 그대로 나타났던 것 같네요 😓 

이번에 구구의 리뷰들을 반영하면서 `외부 라이브러리 의존성을 최대한 격리`시키되, 그러면서도 `내 도메인들의 협력에 구멍이 나지는 않도록`해야겠다는 방향을 가지고 리팩토링을 하게 되었습니다. 그러면서 `도메인 - 라이브러리`의 관계를 `DB Connection - 외부 DB`에 빗대어 생각해보게 되었어요.
도메인에서는 어떤 기능이 필요한지를 추상적으로 정의해서 가지고 있고, 외부 Library는 그 구현체의 형태로 도메인에게 주입되는 것이 적절하겠다는 결론을 내리고 코드에 반영해보았습니다!

KIMSIYOUNG replied:
좋은 결론인 것 같습니다🙂

KIMSIYOUNG commented on src/main/java/wooteco/subway/domain/section/DistanceFare.java:
```diff
@@ -0,0 +1,42 @@
+
+    static Function<Double, Fare> fareCalculator() {
+        return (distance) -> {
+            int totalAmount = BASIC_FARE + OVER_TEN_KM.calculate(distance) + OVER_FIFTY_KM.calculate(distance);
```
현재의 구조에서 80이상의 거리가 추가되면 코드는 어떻게 변경되나요? 새로운 Enum Type을 추가할 순 있지만 계산식은 변경되지 않게 바꿀 수 있을 것 같아요. 

https://github.com/woowacourse/atdd-subway-path/pull/247#discussion_r878782364

KIMSIYOUNG commented on src/main/java/wooteco/subway/domain/path/PathEdge.java:
```diff
@@ -0,0 +1,22 @@
+    }
+
+    @Override
+    protected double getWeight() {
```
라이브러리를 잘 이해하고 구현해주셨네요👏 해당 클래스도 도메인패키지에 있는 클래스가 jgrapht를 의존하고 있네요.

KIMSIYOUNG commented on src/main/java/wooteco/subway/domain/path/AgeFare.java:
```diff
@@ -0,0 +1,35 @@
+        AgeFare applicableAgeFare = Arrays.stream(values())
+                .filter(ageFare -> ageFare.condition.test(age))
+                .findAny()
+                .orElse(OTHER);
```
개인적인 생각입니다.

혹시라도, 위 케이스가 모든 경우의 수를 커버하지 못하는 경우에 OTHER(0원)을 반환하는게 적절할까요?

미션과는 무관하지만 실수로 버그가 발생했을 때 예상하지 못한 엣지케이스에는 전부 0원으로 운임을 메기는 경우가 있을 것 같아요.

이 경우는 에러를 뱉는게 조금 더 적절하지 않을까 싶어요!

YJGwon replied:
여기서 OTHER가 의미하는 것은 할인 받지 않는 일반인을 뜻합니다! field에서 가지는 0이라는 숫자는 1. 할인 요금이 0원 2.할인율이 0원 임을 의미해요. 검증 후 포장된 객체인 `Age`를 다루기 때문에 예상치 못한 값이 들어올 가능성을 막았다고 생각합니다.
그러나 버그라는 것은 정말 어디서 터질지 모르고 100프로 안심할 수 있는 로직이라는 것은 없기 때문에... OTHER에게도 `19세 이상 65세 미만` 조건식을 작성해 준 다음에 `orElseThrow()`로 예외를 발생시키도록 수정해보았어요! 그리고 OTHER라는 네이밍이 코드를 읽는 사람에게 오해를 줄 수 있다는 것을 알게되어서 네이밍도 수정했습니다. 좋은 의견 감사합니다 👍

KIMSIYOUNG replied:
넵넵! 말씀하신대로 ~~하기에 버그가 발생하지 않는다라는 것에서 ~~~하기에 라는 부분을 알지 않아도 버그가 발생하지 않음을 표현할 수 있다면 더 좋을 것 같습니다🙂

KIMSIYOUNG commented on src/main/java/wooteco/subway/domain/section/DistanceFare.java:
```diff
@@ -0,0 +1,42 @@
+        };
+    }
+
+    protected abstract int calculate(double distance);
```
오버라이딩을 통해 구현하는 것과 람다를 통해 구현하는 것이 섞여있는데 어떨 때 다르게 사용하시는 건가요?

YJGwon replied:
- 총 요금 계산의 경우 `Distance`의 값을 꺼낼 필요 없도록 `DistanceFare`에서는 요금 계산 전략만을 전달하도록 만들고 싶어서 Functional Interface를 활용했습니다.
- calculate()의 경우에는 field로 Functional Interface를 사용할 수도 있었을 것 같아요! 현재는 요금 계산 로직을 재귀를 활용해서 구현했기 때문에 method로 선언 해주었어요.

KIMSIYOUNG commented on src/main/java/wooteco/subway/domain/path/Path.java:
```diff
@@ -0,0 +1,73 @@
+
+    private static void checkPath(GraphPath<Station, ? extends DefaultWeightedEdge> path) {
+        if (Objects.isNull(path)) {
+            throw new IllegalStateException("해당하는 경로가 존재하지 않습니다.");
```
이 부분에 대한 테스트도 추가해주시면 좋을 것 같습니다.

KIMSIYOUNG replied:
추가로 여기는 왜 `IllegalStateException` 인가요? `IllegalArgumentException`랑 어떻게 구분하여서 사용하고 계신지 공유해주시면 좋을 것 같아요.

YJGwon replied:
저는 `IllegalArgumentException`은 호출한 쪽 잘못일 때, `IllegalStateException`은 호출한 쪽에서는 유효한 값을 제공했으나 행동하는 객체 측에 문제가 있을 때 사용하고 있어요. 이 경우, 존재하는 출발역과 도착역을 겹치지 않게 잘 입력했는데 경로가 이어져있지 않은 경우기 때문에 **호출하는 쪽은 잘못이 없는데 부득이하게 경로를 찾아줄 수 없는 경우**라고 판단하고 `IllegalStateException`을 사용했습니다.

KIMSIYOUNG replied:
네네🙂 좋은 관점인 것 같습니다.

KIMSIYOUNG commented on src/main/java/wooteco/subway/dto/request/PathRequest.java:
```diff
@@ -0,0 +1,25 @@
+package wooteco.subway.dto.request;
+
+public class PathRequest {
+    private final Long source;
```
Long타입으로 잘 변경해주셨네요🙂 벨리데이션도 같이 추가해주시면 좋을 것 같아요.

YJGwon replied:
해당 field에서 받아오는 Station Id는 Dao에서 예외 처리가 이루어지고 있는데, 말씀하시는 벨리데이션이란 Request Dto 선에서 추가적인 검증을 하는 것을 의미하신걸까요?

KIMSIYOUNG replied:
네네 널이 들어올 수 없다거나 하는 등의 스펙을 작성하면 어떨까 라는 의미였습니다. 사람마다 다르겠지만 컨트롤러를 보고 해당 API의 스펙을 어느정도 유추할 수 있으면 좋다고 생각하는데요. 어떤 값들이 필수값인지 정도에 대한 검증은 앞단에서 하는게 어떨까 라는 의미였습니다.

KIMSIYOUNG commented on src/main/java/wooteco/subway/dto/request/PathRequest.java:
```diff
@@ -0,0 +1,25 @@
+public class PathRequest {
+    private final Long source;
+    private final Long target;
+    private final int age;
```
사용자가 만약 실수로 age값을 입력하지 않은 경우 발생하는 에러에 대한 처리도 같이 해주시면 좋을 것 같아요. 그리고 현재는 바인딩이 되지 않아서 예외가 발생하지만 기본값인 0이 바인딩되면 어떻게 될까요?

YJGwon replied:
아하 Request Field에서 검증할수 있는 최소 검증을 추가하는 것이 맞았군요!! 값에 대한 자세한 검증은 Domain에서 할 수 있게 두되, 입력값이 아예 없는 경우는 Dto선에서 검증하도록 추가하겠습니다.
기본값 바인딩은 어떤 경우에 일어나게 되는 것인지 추측이 안되네요 😓 view쪽에서 기본값을 입력해주었을 경우일까요? 그렇다면 그것까지 막을 수는 없을 것 같아요. 만나이 기준이라 0세 유아를 동반한 승객도 존재할테니까요..!

KIMSIYOUNG replied:
현재는 발생하지 않지만 사용자가 age라는 값을 누락하는 경우(프론트에서 age값을 실수로 넣지 않은 경우)에 누락될 수 있을 것 같아요. 🙂 현재 발생하지 않지만 언제나 값이 누락될 수 있다는 가정이 있는게 안전한 프로그램이라고 생각해서 말씀드렸습니다.

KIMSIYOUNG commented on src/main/java/wooteco/subway/domain/path/Path.java:
```diff
@@ -0,0 +1,73 @@
+package wooteco.subway.domain.path;
```
포키가 첫 번쨰로 질문 주신 부분에 대해서 여기서 같이 얘기해보면 좋을 것 같아요. 

포키가 하고 있는 고민이 저는 좋은 방향이라고 생각해요. X라는 값을 통해서 Y라는 값을 만들어낼 수 있다고해서 `항상 Y는 X에 종속된다 혹은 하위도메인이다` 보긴 어려울 것 같아요. 말씀하신대로 우리가 해결해야하는 영역 또는 문제라는 개념에서 서로 다른 문제를 바라보고 있다면 그 둘은 동일한 데이터를 베이스로 한다고 하더라도 분리할 수 있을 것 같아요. 

제가 미션을 구현 동일하게 경로라는 개념을 새로 도출하고 기존에 라인과 섹션이 관심이 있던 것과 다른 영역으로 간주할 것 같습니다.

KIMSIYOUNG commented on src/main/java/wooteco/subway/ui/service/PathService.java:
```diff
@@ -23,18 +30,20 @@ public PathService(SectionDao sectionDao, StationDao stationDao) {
     }
 
-    public PathResponse getPath(int source, int target) {
+    public PathResponse getPath(PathRequest pathRequest) {
```
`서비스 -> Dao VS 서비스 -> 다른 서비스` 요 부분에 대한 생각은 이미 크루들끼리 많은 얘기와 논쟁이 오가는 것으로 보여요. 

말씀하신대로 경험이 부족하거나 이해가 되지 않는 영역이 있다면 현재까지 이해한 내용에서, dao를 직접의존해도 된다라는 방향을 그대로 밀고가시면 될 것 같아요. 조금 더 다른 것들을 개발하면서도 계속 고민해보는게 더 좋을 것 같습니다.

조금 더 일반화해서 `서비스 > 서비스`, `서비스 > 다오` 가 아니라 시스템과 시스템이라면? 객체와 객체라면? 이라는 관점으로 고민해봐도 괜찮을 것 같아요.

KIMSIYOUNG reviewed (changes requested):
안녕하세요 포키! 
저도 몇가지 코멘트 남겼으니 확인 부탁드립니다🙂

YJGwon committed:
refactor(Jdbc*Dao): DB 저장 후 id값 새 객체에 담아서 return

YJGwon committed:
refactor(Jdbc*Dao): SELECT문에 사용된 `*` 제거

YJGwon committed:
test(AgeTest): 값 비교 method test 추가

YJGwon committed:
test(AgeTest): 값 비교 method test 추가

YJGwon committed:
refactor(Fare): equals 추가

- compareTo()는 equals()와 함께 구현되도록 권고됨: Comparable한 class는 값이 자연적인 순서를 가지며 그 자체로 비교됨을 의미, 따라서 동등성 비교에서도 같은 결과를 얻을 수 있어야

YJGwon committed:
refactor(domain/path): PathAlgorithm 추상화로 `jgrapht` 의존성 격리

YJGwon committed:
refactor(support): jgrapht 관련 객체 package 분리

YJGwon committed:
refactor(DistanceFare): 총 요금 계산 시 stream 활용

- 후에 요금 정책이 더 추가되더라도 총 요금 계산 로직은 수정될 필요 없도록

YJGwon committed:
refactor(AgeFare): 할인받지 않는 일반인도 조건식을 통해 검사

YJGwon committed:
test(PathAlgorithmTest): 경로가 존재하지 않는 case 추가

YJGwon committed:
refactor(dto/request): 요청 validation 추가

YJGwon commented:
안녕하세요!
질문주신 부분들은 제 생각을 comment와 코드를 통해 최대한 표현해보았습니다. 구구와 카일 모두 많은 의견을 나누어주셔서 감사해요! 덕분에 저의 짧았던 고민들을 조금 더 이어나가볼 수 있었습니다.

KIMSIYOUNG reviewed (approved):
안녕하세요 포키!
리뷰 잘 반영해주셨네요🙂
미션은 여기서 머지하도록 하겠습니다.
궁금하신 내용은 편하게 말씀해주세요!
